tests/testthat/test-client-cloud.R

```diff
@@ -0,0 +1,577 @@
+  mockServer = mockServerFactory(list(
+    "^GET /v1/applications/([0-9]+)" = list(
+      content = function(methodAndPath, match, ...) {
+        end <- attr(match, 'match.length')[2] + match[2]
+        application_id <- strtoi(substr(methodAndPath, match[2], end))
+        list(
+          "id"=application_id,
+          "content_id"=application_id-1,
+          "name"=paste("testthat app", application_id)
+        )
+      }
+    ),
+    "^GET /v1/outputs/([0-9]+)" = list(
+      content = function(methodAndPath, match, ...) {
+        end <- attr(match, 'match.length')[2] + match[2]
+        output_id <- strtoi(substr(methodAndPath, match[2], end))
+
+        list(
+          "id"=output_id,
+          "source_id"=output_id + 1,
+          "url"="http://fake-url.test.me/"
+        )
+      }
+    )
+  ))
+
+  restoreOpt <- options(rsconnect.http = mockServer)
+  withr::defer(options(restoreOpt))
+
+  testAccount <- configureTestAccount()
+  withr::defer(removeAccount(testAccount))
+
+  target <- deploymentTargetForApp(
+    appId = 3,
+    account = testAccount,
+    server = 'posit.cloud',
+  )
+
+  expect_equal(target$appName, "testthat app 3")
+  expect_equal(target$account, testAccount)
+  expect_equal(target$server, 'posit.cloud')
+  expect_equal(target$appId, 3)
+})
+
+deployAppMockServerFactory <- function(expectedAppType) {
+  outputResponse <- list(
+    "id"=1,
+    "source_id"=2,
+    "url"="http://fake-url.test.me/"
+  )
+
+  actualCalls = list(
+    outputCreated = FALSE,
+    revisionCreated = FALSE,
+    bundleCreated = FALSE,
+    bundleUploaded = FALSE,
+    bundleReady = FALSE,
+    deployStarted = FALSE
+  )
+
+  server <- mockServerFactory(list(
+    # An attempt to search for preexisting applications with the same name.
+    # This call should change, see https://github.com/rstudio/rsconnect/issues/808
+    "^GET /v1/applications/[?]filter=account_id:50&filter=type:connect"=list(
+      content=list(
+        count=0,
+        total=0,
+        applications=list()
+      )
+    ),
+    "^POST /v1/outputs$"=list(
+      content=function(methodAndPath, match, contentFile, ...) {
+        e <- environment(); p <- parent.env(e); p$actualCalls$outputCreated <- TRUE
+
+        content = jsonlite::fromJSON(readChar(contentFile, file.info(contentFile)$size))
+
+        expect_equal(content$application_type, expectedAppType)
+        expect_equal(content$name, "Desired name here")
+
+        outputResponse
+      }
+    ),
+    "^GET /v1/applications/([0-9]+)$" = list(
+      content = function(methodAndPath, match, ...) {
+        end <- attr(match, 'match.length')[2] + match[2]
+        application_id <- strtoi(substr(methodAndPath, match[2], end))
+
+        list(
+          id=application_id,
+          content_id=application_id-1,
+          name=paste("testthat app", application_id),
+          type=expectedAppType
+        )
+      }
+    ),
+    "^POST /v1/bundles/123/status$" = list(
```

Comment:
Is this supposed to be a GET instead of a POST?

Reply:
No, this is a real call we expect rsconnect to make. It's like 1. Create bundle, status is pending; 2. Upload bundle to S3; 3. Let lucid-server know we've done that by making the call mocked here.